handlers/ecs/textextraction/s3handler.py

```diff
@@ -0,0 +1,188 @@
+        # posixpath.normpath() strips the trailing /. Add it back.
+        if path.endswith('/') or _final_path + '/' == final_path:
+            _final_path += '/'
+        final_path = _final_path
+    if final_path == base_path:
+        final_path += '/'
+    base_path_len = len(base_path)
+    if (not final_path.startswith(base_path) or final_path[base_path_len] != '/'):
+        raise ValueError('the joined path is located outside of the base path component')
+    return final_path.lstrip('/')
+
+
+class Storage():
+    DEFAULT_CONTENT_TYPE = 'application/octet-stream'
+    DEFAULT_QUERYSTRING_EXPIRE = 3600  # 1 hour
+
+    QUERYSTRING_AUTH = True
+    GZIP_CONTENT_TYPES = [
+        'text/javascript', 'application/javascript', 'application/x-javascript',
+        'image/svg+xml', 'application/json', 'text/css',
+    ]
+
+    def __init__(self, bucket_name, base_path):
+        self.bucket_name = bucket_name
+        self.base_path = base_path
+        self._file = None
+
+    @property
+    def connection(self):
+        connection = getattr(self, '_connection', None)
+        if connection is None:
+            session = boto3.session.Session()
+            self._connection = session.resource('s3')
+        return self._connection
+
+    @property
+    def bucket(self):
+        bucket = getattr(self, '_bucket', None)
+        if bucket is None:
+            self._bucket = self.connection.Bucket(self.bucket_name)
+        return self._bucket
+
+    def exists(self, name):
+        try:
+            self.connection.meta.client.head_object(Bucket=self.bucket_name, Key=name)
+            return True
+        except ClientError:
+            return False
+
+    def get_alternative_name(self, file_root, file_ext):
+        return '%s_%s%s' % (file_root, get_random_string(7), file_ext)
+
+    def get_available_name(self, name, max_length=1024):
+        dir_name, file_name = os.path.split(name)
+        file_root, file_ext = os.path.splitext(file_name)
+        while self.exists(name) or (max_length and len(name) > max_length):
+            name = os.path.join(dir_name, self.get_alternative_name(file_root, file_ext))
+            if max_length is None:
+                continue
+            truncation = len(name) - max_length
+            if truncation > 0:
+                file_root = file_root[:-truncation]
+                if not file_root:
+                    raise Exception(f"Can't generate filename for '{name}' due to s3 key size limit.")
+                name = os.path.join(dir_name, self.get_alternative_name(file_root, file_ext))
+        return name
+
+    def _compress_content(self, content):
+        """Gzip a given string content."""
+        content.seek(0)
+        zbuf = io.BytesIO()
+        zfile = GzipFile(mode='wb', fileobj=zbuf, mtime=0.0)
+        try:
+            zfile.write(force_bytes(content.read()))
+        finally:
+            zfile.close()
```

Comment:
maybe with statement instead of try finally? The closing would automatically be handled.

Reply:
let's keep it like that.